bench/minetest/optimized/cavegen.cpp.ll

```diff
@@ -1999,397 +1999,400 @@ define dso_local void @_ZN15CavesRandomWalk10makeTunnelEb(ptr nocapture noundef
   %14 = trunc i32 %13 to i16
   %15 = and i16 %14, 32767
   %16 = urem i16 %15, 20
-  %17 = uitofp i16 %16 to float
-  %18 = fadd nsz float %17, -1.000000e+01
-  %19 = fdiv nsz float %18, 1.000000e+01
-  %20 = getelementptr inbounds i8, ptr %0, i64 120
-  %21 = getelementptr inbounds i8, ptr %0, i64 128
-  %22 = mul i32 %12, 1103515245
+  %17 = add nsw i16 %16, -10
+  %18 = getelementptr inbounds i8, ptr %0, i64 120
+  %19 = getelementptr inbounds i8, ptr %0, i64 128
+  %20 = mul i32 %12, 1103515245
+  %21 = add i32 %20, 12345
+  %22 = mul i32 %21, 1103515245
   %23 = add i32 %22, 12345
-  %24 = mul i32 %23, 1103515245
-  %25 = add i32 %24, 12345
-  %26 = insertelement <2 x i32> poison, i32 %25, i64 0
-  %27 = insertelement <2 x i32> %26, i32 %23, i64 1
-  %28 = sdiv <2 x i32> %27, <i32 65536, i32 65536>
-  %29 = trunc <2 x i32> %28 to <2 x i16>
-  %30 = and <2 x i16> %29, <i16 32767, i16 32767>
-  %31 = urem <2 x i16> %30, <i16 20, i16 20>
-  %32 = uitofp <2 x i16> %31 to <2 x float>
-  %33 = fadd nsz <2 x float> %32, <float -1.000000e+01, float -1.000000e+01>
-  %34 = fdiv nsz <2 x float> %33, <float 1.000000e+01, float 3.000000e+01>
-  %35 = mul i32 %25, 1103515245
-  %36 = add i32 %35, 12345
-  store i32 %36, ptr %8, align 4, !tbaa !147
-  %37 = sdiv i32 %36, 65536
-  %38 = trunc i32 %37 to i16
-  %39 = and i16 %38, 32767
-  %40 = urem i16 %39, 11
-  %41 = uitofp i16 %40 to float
-  %42 = fdiv nsz float %41, 1.000000e+01
-  %43 = insertelement <2 x float> poison, float %42, i64 0
-  %44 = shufflevector <2 x float> %43, <2 x float> poison, <2 x i32> zeroinitializer
-  %45 = fmul nsz <2 x float> %34, %44
-  store <2 x float> %45, ptr %20, align 8, !tbaa !70
-  %46 = fmul nsz float %19, %42
-  store float %46, ptr %21, align 8, !tbaa !130
-  br label %47
+  %24 = insertelement <2 x i32> poison, i32 %23, i64 0
+  %25 = insertelement <2 x i32> %24, i32 %21, i64 1
+  %26 = sdiv <2 x i32> %25, <i32 65536, i32 65536>
+  %27 = trunc <2 x i32> %26 to <2 x i16>
+  %28 = and <2 x i16> %27, <i16 32767, i16 32767>
+  %29 = urem <2 x i16> %28, <i16 20, i16 20>
+  %30 = add nsw <2 x i16> %29, <i16 -10, i16 -10>
+  %31 = sitofp <2 x i16> %30 to <2 x float>
+  %32 = fdiv nsz <2 x float> %31, <float 1.000000e+01, float 3.000000e+01>
+  %33 = mul i32 %23, 1103515245
+  %34 = add i32 %33, 12345
+  store i32 %34, ptr %8, align 4, !tbaa !147
+  %35 = sdiv i32 %34, 65536
+  %36 = trunc i32 %35 to i16
+  %37 = and i16 %36, 32767
+  %38 = urem i16 %37, 11
+  %39 = insertelement <2 x i16> poison, i16 %17, i64 0
+  %40 = insertelement <2 x i16> %39, i16 %38, i64 1
+  %41 = sitofp <2 x i16> %40 to <2 x float>
+  %42 = uitofp <2 x i16> %40 to <2 x float>
+  %43 = shufflevector <2 x float> %41, <2 x float> %42, <2 x i32> <i32 0, i32 3>
+  %44 = fdiv nsz <2 x float> %43, <float 1.000000e+01, float 1.000000e+01>
```

Comment:
Regression?

Reply:
See my commends on the PR itself, but I don't think this is a regression. It seems to be profitable as it saves a division.

Reply:
Oh, I see. I am just curious about the IR:
```
%39 = insertelement <2 x i16> poison, i16 %17, i64 0
%40 = insertelement <2 x i16> %39, i16 %38, i64 1
%41 = sitofp <2 x i16> %40 to <2 x float>
%42 = uitofp <2 x i16> %40 to <2 x float>
%43 = shufflevector <2 x float> %41, <2 x float> %42, <2 x i32> <i32 0, i32 3>
``` 
It can be folded into:
```
%a = sitofp i16 %17 to float
%b = uitofp i16 %38 to float
%39 = insertelement <2 x float> poison, float %a, i64 0
%40 = insertelement <2 x float> %39, float %b, i64 1
```

Reply:
Yeah that seems like a mistake. Think its better to try and fixup in SLP, or just make a fold in InstCombine?

Reply:
Working on a patch in InstCombine.

Reply:
Just a thought, but may in your script for generating the optimized ll files, you could dump (or aggregate) `instcount` or `print<cost-model>`. Might be finding regressions a bit more automated.

Reply:
> Oh, I see. I am just curious about the IR:
> 
> ```
> %39 = insertelement <2 x i16> poison, i16 %17, i64 0
> %40 = insertelement <2 x i16> %39, i16 %38, i64 1
> %41 = sitofp <2 x i16> %40 to <2 x float>
> %42 = uitofp <2 x i16> %40 to <2 x float>
> %43 = shufflevector <2 x float> %41, <2 x float> %42, <2 x i32> <i32 0, i32 3>
> ```
> 
> It can be folded into:
> 
> ```
> %a = sitofp i16 %17 to float
> %b = uitofp i16 %38 to float
> %39 = insertelement <2 x float> poison, float %a, i64 0
> %40 = insertelement <2 x float> %39, float %b, i64 1
> ```

See: https://github.com/llvm/llvm-project/pull/84645

Reply:
> Just a thought, but may in your script for generating the optimized ll files, you could dump (or aggregate) `instcount` or `print<cost-model>`. Might be finding regressions a bit more automated.

see: https://github.com/dtcxzyw/llvm-opt-benchmark/pull/355